ProjectProtocol/project-protocol-web
# [PRO-264] Support data caching with react-query

The problem
---
If a user scrolls down past the first page of results, and clicks on a search result, then clicks back to return to the list, they will not be returned to the same position in the list.

This is due to the search data not being cached so when the user clicks back, the request has to refire the API request and rebuild the list, but it is unable to return to the exact position of the clicked-on agent as expected. To reproduce this, you can visit https://ratemypo-staging.netlify.app/, scroll down past the first page, click on an agent, click back, and see that the screen doesn't reposition itself properly.

We've been using `react-router` loaders to load our data, but they seem increasingly brittle and hard to make work. They have no access to react component lifecycles or shared state, because they exist outside of react components and are called before render.

React-Query is a very popular library, and has caching and pagination functionality (and a ton of other stuff) right out of the box.

It basically makes our `Paginator` component unnecessary, and handles the "get more", and data appending logic for us, all while caching results. I think `Search.tsx` in this branch demonstrates how react-query gives us a slightly more simplified approach.

One thing to note is that instead of one long list of items, react-query keeps an array of pages. To render, we have to map through the pages, and then map through each item of each page.

Changes
---
* Add [@tanstack/react-query](https://tanstack.com/query/latest/docs/framework/react/overview) - seems to be a go-to tool in the react community and has tons of functionality built in. It supports lots of simple and high-complexity use-cases.
* Remove search loader from search route
* Wrap each page of results in `AnimatedList`
* Attempt to smooth out AnimatedList rendering by preventing animation for items that aren't on the screen.
  * Animation is a work in progress but I think is good enough for now.
* Add `useLoadingBar` hook so we can trigger the loading indicator wherever/whenever needed.

To test
---
* Please have a click around the deploy preview, do a few searches, 
* Test going back and forth between the search list, clicking on an agent, and going back. You should be returned to the same place on the list.
